[JIT] Introduce BuiltinOpFunction and integrate into torchbind

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34140 [WIP][quantization] Use torchbind for Linear PackedParams
* #34339 [quantization] Make FP16 RNN use new prepack op
* **#34098 [JIT] Introduce BuiltinOpFunction and integrate into torchbind**
* #33921 [JIT] Virtualize Function
* #33900 [JIT] Move stuff out of class_type.cpp

Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

Commit:
[JIT] Introduce BuiltinOpFunction and integrate into torchbind

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/builtin_function.h:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  void run(Stack& stack) override {
+    auto handle = c10::Dispatcher::singleton().findSchemaOrThrow(
```
BuiltinOp should just store its Schema, and std::function<void(Stack&)> directly. It does not have to participate in dispatch (and in fact will mostly only incur overhead because of it).

dzhulgakov replied:
@zdevito - hm, are you sure about that? I'd at least want to have it go through dispatcher to get profiling working. TorchBind ops can be potentially big so it's good to capture them through the same profiling infra. Alternatively - we can just slap the same RecordFunction here too

Also I'm pretty sure someone will want tracing to work too (though it's debatable)

zdevito commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1825,6 +1826,19 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
 
 std::vector<Value*> inlineCallTo(Node* to_replace, Function* callee) {
   WithInsertPoint guard(to_replace);
+  if (!callee->isGraphFunction()) {
```
I think the right place to edit is inliner.cpp not the call inlineCallTo. This should just fail for a builtin (since it is builtin it cannot be inlined).

zdevito commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1147,6 +1147,9 @@ struct PythonPrintImpl {
   void printFunction(
       const Function& func,
       bool print_first_argument_type = true) {
+    if (!func.isGraphFunction()) {
```
Serialization will need to unify builtin classes with their corresponding class in the .pt file.  We haven't discussed the expected behavior, but I think we should go further and  _not serialize_ a class that contains builtin methods. We should consider it to be something that can be looked up when read in. Otherwise we get into tricky situations where there is a mix between the already loaded builtin class and what is in the serialized file.

zdevito commented on torch/custom_class.h:
```diff
@@ -163,40 +170,15 @@ class class_ {
-
-    auto method = classCU()->create_function(qualClassName + "." + name, graph);
-    classTypePtr->addMethod(method);
+
```
Rather than register an operator here, use the boxing facilities of c10 to generate a `std::function<void(Stack&)>` and then use that to directly construct a BuiltinOpFunction.

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"




[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"




[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20226171](https://our.internmc.facebook.com/intern/diff/D20226171)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20226171](https://our.internmc.facebook.com/intern/diff/D20226171)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/builtin_function.h:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  const c10::FunctionSchema& getSchema() const override {
+    if (!adapted_schema_) {
```
technically it's not threadsafe and also you'd need double-locking. But why not be safe and always take the mutex if it's not in the hot path?

jamesr66a replied:
Technically that wouldn't be thread-safe either because we're returning a const ref here

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -58,16 +61,21 @@ class class_ {
+    classTypePtr->addAttribute("capsule", at::CapsuleType::get());
+
+    c10::getCustomClassTypeMap().insert(
+        {typeid(c10::intrusive_ptr<CurClass>).name(),
```
do you need demangle here and everywhere? Also - do we need to rely on RTTI for it? you could also use c10::get_fully_qualified_type_name (not sure where exactly we use these names). Or we never use those names other than for accounting? (in which case maybe we should just use typeid directly)

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -163,40 +171,22 @@ class class_ {
+    auto wrapped_func = [func = std::move(func)](Stack& stack) mutable -> void {
+      using RetType =
+          typename c10::guts::infer_function_traits_t<Func>::return_type;
+      detail::BoxedProxy<RetType, Func>()(stack, func);
```
so the question is how to handle profiling for these ops. If we don't go through dispatcher - maybe we should call it here directly (with something like https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/register_c10_ops.cpp#L18 )

jamesr66a replied:
I'll just add a RECORD_FUNCTION

zdevito replied:
hot take: we should RECORD_FUNCTION in CALL instructions in the interpreter. That would catch this case and also show what time we are spending in actual methods.

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

zdevito commented on torch/custom_class_detail.h:
```diff
@@ -60,10 +62,68 @@ Func wrap_func(Func f) {
+    bool AllowDeprecatedTypes,
+    size_t... ivalue_arg_indices>
+typename c10::guts::infer_function_traits_t<Functor>::return_type
+call_torchbind_method_from_stack(
```
This appears to be unused?

jamesr66a replied:
It's used in BoxedProxy below

zdevito commented on torch/custom_class.h:
```diff
@@ -58,16 +61,21 @@ class class_ {
+             std::shared_ptr<script::CompilationUnit>(), classTypePtr)});
+    c10::getCustomClassTypeMap().insert(
+        {typeid(c10::tagged_capsule<CurClass>).name(),
+         c10::StrongTypePtr(
```
If the class map also has null CompilationUnits, can we just store name -> ClassType directly?

zdevito commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -1,27 +1,49 @@
 
-void setGetCustomClassFn(GetCustomClassFnType fn) {
-  custom_class_fn.store(fn);
+Function* getCustomClassMethod(const std::string& qualname) {
```
This appears unused.

zdevito commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -1,27 +1,49 @@
 
-at::TypePtr getCustomClass(const std::string& name) {
-  return custom_class_fn.load()(name);
+void registerCustomClassMethod(std::shared_ptr<Function> fn) {
```
Can we just stash them in a flat list to keep them alive? I want to avoid giving the impression that someone should look at this table to look up a method. They should go directly to the table in ClassType.

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -367,10 +367,6 @@ StrongTypePtr::StrongTypePtr(
   cu_ = std::move(cu);
   type_ = type;
   TORCH_INTERNAL_ASSERT(type_);
-  if (type_->cast<ClassType>()) {
```
We can possiblly keep this assert if the class table no longer holds StrongTypePtr objects

jamesr66a replied:
`ivalue::object::create` requires a `StrongTypePtr` for instantiation of any object, and if it's a custom class that's going to be nullptr, so I don't think we can remove this

zdevito reviewed (commented):
This looks like it is going to work. Thanks for all the revisions. I just have some minor things to resolve inline.

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1095,13 +1112,11 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
-            frames.back().pc = af.pc + 1;
-            enterFrame(code, stack.size() - inst.N);
-            af = ActiveFrame(frames.back());
+            if (!function->isGraphFunction()) {
```
nit: maybe have runFunction be a function, since this if-statement occurs twice.

Commit:
Update on "[JIT] Introduce BuiltinOpFunction and integrate into torchbind"


Differential Revision: [D20229166](https://our.internmc.facebook.com/intern/diff/D20229166)

[ghstack-poisoned]
